<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_33) on Tue Aug 14 12:30:35 EDT 2012 -->
<TITLE>
InputParser (${Name} API)
</TITLE>

<META NAME="date" CONTENT="2012-08-14">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="InputParser (${Name} API)";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../sanger/margarita/Imputer.html" title="class in sanger.margarita"><B>PREV CLASS</B></A>&nbsp;
&nbsp;NEXT CLASS</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?sanger/margarita/InputParser.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="InputParser.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
sanger.margarita</FONT>
<BR>
Class InputParser</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>sanger.margarita.InputParser</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>InputParser</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#InputParser()">InputParser</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new instance of InputParser</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;short[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#cloneAlleleCounts()">cloneAlleleCounts</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;How many of each allele at each marker?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.LinkedList&lt;int[]&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#clonePossibleMutations()">clonePossibleMutations</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Which mutations are immediately possible?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;byte[][]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#getInputSequences()">getInputSequences</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;What are the input sequences?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#getMarkerLocations()">getMarkerLocations</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;What are the positions of the markers?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#getNumberOfCases()">getNumberOfCases</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;How many case haplotype sequences are there in the data?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#getNumberOfControls()">getNumberOfControls</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;How many control haplotype sequences are there in the data?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#getNumberOfMarkers()">getNumberOfMarkers</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;How many markers are typed in the data?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#getNumberOfSequences()">getNumberOfSequences</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;How many haplotype sequences are there in the data?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#getNumberOfUnphasedCharacters()">getNumberOfUnphasedCharacters</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;How many unphased or missing characters are there in the data?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../sanger/margarita/InputParser.html#parseFile(sanger.argml.environment.Environment, sanger.argml.io.TextInput)">parseFile</A></B>(<A HREF="../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
          <A HREF="../../sanger/argml/io/TextInput.html" title="class in sanger.argml.io">TextInput</A>&nbsp;input)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reads in a file that is in Margarita format.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="InputParser()"><!-- --></A><H3>
InputParser</H3>
<PRE>
public <B>InputParser</B>()</PRE>
<DL>
<DD>Creates a new instance of InputParser
<P>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="parseFile(sanger.argml.environment.Environment, sanger.argml.io.TextInput)"><!-- --></A><H3>
parseFile</H3>
<PRE>
public final void <B>parseFile</B>(<A HREF="../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
                            <A HREF="../../sanger/argml/io/TextInput.html" title="class in sanger.argml.io">TextInput</A>&nbsp;input)</PRE>
<DL>
<DD>Reads in a file that is in Margarita format.
 The data structures are initialised that will be acted upon
 by ArgBuilder.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>filename</CODE> - The file that contains the marker
  locations and sequences in Margarita format.</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfCases()"><!-- --></A><H3>
getNumberOfCases</H3>
<PRE>
public final int <B>getNumberOfCases</B>()</PRE>
<DL>
<DD>How many case haplotype sequences are there in the data?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of case haplotype sequences.</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfControls()"><!-- --></A><H3>
getNumberOfControls</H3>
<PRE>
public final int <B>getNumberOfControls</B>()</PRE>
<DL>
<DD>How many control haplotype sequences are there in the data?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of control haplotype sequences.</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfSequences()"><!-- --></A><H3>
getNumberOfSequences</H3>
<PRE>
public final int <B>getNumberOfSequences</B>()</PRE>
<DL>
<DD>How many haplotype sequences are there in the data?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of haplotype sequences.</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfMarkers()"><!-- --></A><H3>
getNumberOfMarkers</H3>
<PRE>
public final int <B>getNumberOfMarkers</B>()</PRE>
<DL>
<DD>How many markers are typed in the data?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of markers.</DL>
</DD>
</DL>
<HR>

<A NAME="getNumberOfUnphasedCharacters()"><!-- --></A><H3>
getNumberOfUnphasedCharacters</H3>
<PRE>
public final int <B>getNumberOfUnphasedCharacters</B>()</PRE>
<DL>
<DD>How many unphased or missing characters are there in the data?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of unphased or missing characters.</DL>
</DD>
</DL>
<HR>

<A NAME="getMarkerLocations()"><!-- --></A><H3>
getMarkerLocations</H3>
<PRE>
public final double[] <B>getMarkerLocations</B>()</PRE>
<DL>
<DD>What are the positions of the markers?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the positions of the markers.</DL>
</DD>
</DL>
<HR>

<A NAME="getInputSequences()"><!-- --></A><H3>
getInputSequences</H3>
<PRE>
public final byte[][] <B>getInputSequences</B>()</PRE>
<DL>
<DD>What are the input sequences?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the input sequences, where the first dimension is the sequence index,
 and the second dimension is the marker index.
 0 and 1 are the alleles, 2 means unphased and 3 means missing data.</DL>
</DD>
</DL>
<HR>

<A NAME="cloneAlleleCounts()"><!-- --></A><H3>
cloneAlleleCounts</H3>
<PRE>
public final short[][] <B>cloneAlleleCounts</B>()</PRE>
<DL>
<DD>How many of each allele at each marker?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of alleles at each markers. [0][marker] is
 the number of 0 alleles at marker, similarly for [1][marker]
 and [2][marker] is the number of missing data at that marker.</DL>
</DD>
</DL>
<HR>

<A NAME="clonePossibleMutations()"><!-- --></A><H3>
clonePossibleMutations</H3>
<PRE>
public final java.util.LinkedList&lt;int[]&gt; <B>clonePossibleMutations</B>()</PRE>
<DL>
<DD>Which mutations are immediately possible?
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the number of alleles at each markers. [0][marker] is
 the number of 0 alleles at marker, similarly for [1][marker]
 and [2][marker] is the number of missing data at that marker.</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../sanger/margarita/Imputer.html" title="class in sanger.margarita"><B>PREV CLASS</B></A>&nbsp;
&nbsp;NEXT CLASS</FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?sanger/margarita/InputParser.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="InputParser.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>
Copyright (c) 2008 Genome Research Ltd.
</BODY>
</HTML>
