<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_33) on Tue Aug 14 12:30:35 EDT 2012 -->
<TITLE>
Genealogy (${Name} API)
</TITLE>

<META NAME="date" CONTENT="2012-08-14">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Genealogy (${Name} API)";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../sanger/argml/graph/model/Edge.html" title="class in sanger.argml.graph.model"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../sanger/argml/graph/model/GenealogyFactory.html" title="class in sanger.argml.graph.model"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?sanger/argml/graph/model/Genealogy.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Genealogy.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
sanger.argml.graph.model</FONT>
<BR>
Class Genealogy</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><A HREF="../../../../sanger/argml/environment/Environmental.html" title="class in sanger.argml.environment">sanger.argml.environment.Environmental</A>
      <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><B>sanger.argml.graph.model.Genealogy</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.lang.Cloneable</DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>Genealogy</B><DT>extends <A HREF="../../../../sanger/argml/environment/Environmental.html" title="class in sanger.argml.environment">Environmental</A><DT>implements java.lang.Cloneable</DL>
</PRE>

<P>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#ancestralSequence">ancestralSequence</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Edge.html" title="class in sanger.argml.graph.model">Edge</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#edges">edges</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#gmrca">gmrca</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#haplotypeDomain">haplotypeDomain</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#snpDomain">snpDomain</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#vertices">vertices</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#Genealogy(sanger.argml.environment.Environment, sanger.math.set.NaturalDomain, sanger.math.set.NaturalDomain)">Genealogy</A></B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
          <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;snpDomain,
          <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;haplotypeDomain)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Construct an empty genealogy and populates a vertex for every haplotype.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected </CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#Genealogy(sanger.argml.environment.Environment, sanger.math.set.NaturalDomain, sanger.math.set.NaturalDomain, sanger.math.set.NaturalSet)">Genealogy</A></B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
          <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;snpDomain,
          <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;haplotypeDomain,
          <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;ancestral)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Construct an empty genealogy.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#ancestralSequence()">ancestralSequence</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#biPartitions()">biPartitions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculate all possible bipartitions for the tree.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#clipSnpDomain(sanger.argml.graph.model.Genealogy, sanger.math.set.NaturalSet)">clipSnpDomain</A></B>(<A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A>&nbsp;genealogy,
              <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Clip the SNP domain.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#clipSnpDomain(sanger.argml.graph.model.Genealogy, sanger.math.set.NaturalSet, sanger.argml.graph.model.Vertex)">clipSnpDomain</A></B>(<A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A>&nbsp;genealogy,
              <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region,
              <A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A>&nbsp;vertex)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Clip the SNP domain from <code>vertex</code> to the present.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#clone()">clone</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Edge.html" title="class in sanger.argml.graph.model">Edge</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#edges()">edges</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#gmrca()">gmrca</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#haplotypeDomain()">haplotypeDomain</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#haplotypes()">haplotypes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reconstruct the haplotypes on all the leaves by tracing mutation events on the lineage.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#haplotypes(sanger.math.set.NaturalSet)">haplotypes</A></B>(<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Reconstruct the haplotypes in <code>region</code> on all the leaves by tracing mutation events on the lineage.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#isTree()">isTree</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#leaves()">leaves</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Return a list of vertices with zero <A HREF="../../../../sanger/argml/graph/model/Vertex.html#outDegree()"><CODE>out degree</CODE></A>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.HashMap&lt;<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>,<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#localTreesBiPartitions()">localTreesBiPartitions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculate all possible bipartitions for all local trees in the graph.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#mutations()">mutations</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trace all mutation events on the lineage leading to all the leaves.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#mutations(sanger.math.set.NaturalSet)">mutations</A></B>(<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Trace all mutation events in <code>region</code> on the lineage leading to all the leaves.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#print(sanger.argml.io.TextOutput)">print</A></B>(<A HREF="../../../../sanger/argml/io/TextOutput.html" title="class in sanger.argml.io">TextOutput</A>&nbsp;printer)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#printDot1(sanger.argml.io.TextOutput)">printDot1</A></B>(<A HREF="../../../../sanger/argml/io/TextOutput.html" title="class in sanger.argml.io">TextOutput</A>&nbsp;printer)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#printDot2(sanger.argml.io.TextOutput)">printDot2</A></B>(<A HREF="../../../../sanger/argml/io/TextOutput.html" title="class in sanger.argml.io">TextOutput</A>&nbsp;printer)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#recombinationFreeRegions()">recombinationFreeRegions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A collection of non overlapping sets where no recombination events occur.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#recombinationPositions()">recombinationPositions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Map locations where recombination events occur.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#recombinationRates()">recombinationRates</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Count the recombinations on the graph for every snp.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#reportEmptyRegions()">reportEmptyRegions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Report edges with empty active region to enviroment log.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#setAncestralSequence(sanger.math.set.NaturalSet)">setAncestralSequence</A></B>(<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;ancestralSequence)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#snpDomain()">snpDomain</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#sort()">sort</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sort the vertex and edge collections according to their natural order.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/Genealogy.html#vertices()">vertices</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_sanger.argml.environment.Environmental"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class sanger.argml.environment.<A HREF="../../../../sanger/argml/environment/Environmental.html" title="class in sanger.argml.environment">Environmental</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../sanger/argml/environment/Environmental.html#env()">env</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="snpDomain"><!-- --></A><H3>
snpDomain</H3>
<PRE>
protected <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A> <B>snpDomain</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="haplotypeDomain"><!-- --></A><H3>
haplotypeDomain</H3>
<PRE>
protected <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A> <B>haplotypeDomain</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="ancestralSequence"><!-- --></A><H3>
ancestralSequence</H3>
<PRE>
protected <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A> <B>ancestralSequence</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="vertices"><!-- --></A><H3>
vertices</H3>
<PRE>
protected java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A>&gt; <B>vertices</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="edges"><!-- --></A><H3>
edges</H3>
<PRE>
protected java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Edge.html" title="class in sanger.argml.graph.model">Edge</A>&gt; <B>edges</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="gmrca"><!-- --></A><H3>
gmrca</H3>
<PRE>
protected <A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A> <B>gmrca</B></PRE>
<DL>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="Genealogy(sanger.argml.environment.Environment, sanger.math.set.NaturalDomain, sanger.math.set.NaturalDomain)"><!-- --></A><H3>
Genealogy</H3>
<PRE>
public <B>Genealogy</B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
                 <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;snpDomain,
                 <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;haplotypeDomain)</PRE>
<DL>
<DD>Construct an empty genealogy and populates a vertex for every haplotype.
 The ancestral sequence is set to 0's.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>env</CODE> - The runtime enviroment<DD><CODE>snpDomain</CODE> - The snp domain of the haplotypes in the sample.<DD><CODE>haplotypeDomain</CODE> - The haplotype domain of the sample.</DL>
</DL>
<HR>

<A NAME="Genealogy(sanger.argml.environment.Environment, sanger.math.set.NaturalDomain, sanger.math.set.NaturalDomain, sanger.math.set.NaturalSet)"><!-- --></A><H3>
Genealogy</H3>
<PRE>
protected <B>Genealogy</B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
                    <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;snpDomain,
                    <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;haplotypeDomain,
                    <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;ancestral)</PRE>
<DL>
<DD>Construct an empty genealogy.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>env</CODE> - The runtime enviroment<DD><CODE>snpDomain</CODE> - The snp domain of the haplotypes in the sample.<DD><CODE>haplotypeDomain</CODE> - The haplotype domain of the sample.<DD><CODE>ancestral</CODE> - Ancestral sequence on the grand most recent common ancestor.</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="biPartitions()"><!-- --></A><H3>
biPartitions</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A> <B>biPartitions</B>()</PRE>
<DL>
<DD>Calculate all possible bipartitions for the tree.
 This procedure applies only when <A HREF="../../../../sanger/argml/graph/model/Genealogy.html#isTree()"><CODE>isTree()</CODE></A> return true.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>A collection of all possible bipartitions.</DL>
</DD>
</DL>
<HR>

<A NAME="localTreesBiPartitions()"><!-- --></A><H3>
localTreesBiPartitions</H3>
<PRE>
public java.util.HashMap&lt;<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>,<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A>&gt; <B>localTreesBiPartitions</B>()</PRE>
<DL>
<DD>Calculate all possible bipartitions for all local trees in the graph.
 The map stores a collection of bipartions for the local tree for every <A HREF="../../../../sanger/argml/graph/model/Genealogy.html#recombinationFreeRegions()"><CODE>recombination free regions</CODE></A> in the snp domain.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>A map of collections of bipartitions.</DL>
</DD>
</DL>
<HR>

<A NAME="recombinationRates()"><!-- --></A><H3>
recombinationRates</H3>
<PRE>
public int[] <B>recombinationRates</B>()</PRE>
<DL>
<DD>Count the recombinations on the graph for every snp.
 The array corresponds to the <A HREF="../../../../sanger/argml/graph/model/Genealogy.html#snpDomain()"><CODE>snp domain's</CODE></A> relative coordinates,
 use <A HREF="../../../../sanger/math/set/NaturalDomain.html#toAbsoluteCoordinate(int)"><CODE>NaturalDomain.toAbsoluteCoordinate(int)</CODE></A> to get absolute position in snp domain.
 A value in cell i counts the recombinations that occur between the position of snp i and i+1.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>An array of integers with recombination count for every snp in the domain.</DL>
</DD>
</DL>
<HR>

<A NAME="recombinationPositions()"><!-- --></A><H3>
recombinationPositions</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A> <B>recombinationPositions</B>()</PRE>
<DL>
<DD>Map locations where recombination events occur.
 Use <A HREF="../../../../sanger/argml/graph/model/Genealogy.html#recombinationRates()"><CODE>recombinationRates()</CODE></A> for actual recombination counts.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>A set of all positions where recombination events occur.</DL>
</DD>
</DL>
<HR>

<A NAME="recombinationFreeRegions()"><!-- --></A><H3>
recombinationFreeRegions</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A> <B>recombinationFreeRegions</B>()</PRE>
<DL>
<DD>A collection of non overlapping sets where no recombination events occur.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>A collection of sets from the snp domain.</DL>
</DD>
</DL>
<HR>

<A NAME="clipSnpDomain(sanger.argml.graph.model.Genealogy, sanger.math.set.NaturalSet)"><!-- --></A><H3>
clipSnpDomain</H3>
<PRE>
public static <A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A> <B>clipSnpDomain</B>(<A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A>&nbsp;genealogy,
                                      <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region)
                               throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD>Clip the SNP domain.
 Constructs a new genealogy spanning only the positions contained in <code>region</code>.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>genealogy</CODE> - Genealogy to clip <code>region</code> from.<DD><CODE>region</CODE> - Set of position to crop.
<DT><B>Returns:</B><DD>A new genealogy containing the clip.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE> - if <code>region</code> is not from the genealogy's snp domain.</DL>
</DD>
</DL>
<HR>

<A NAME="clipSnpDomain(sanger.argml.graph.model.Genealogy, sanger.math.set.NaturalSet, sanger.argml.graph.model.Vertex)"><!-- --></A><H3>
clipSnpDomain</H3>
<PRE>
public static <A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A> <B>clipSnpDomain</B>(<A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A>&nbsp;genealogy,
                                      <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region,
                                      <A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A>&nbsp;vertex)
                               throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD>Clip the SNP domain from <code>vertex</code> to the present.
 Constructs a new genealogy from the sub graph rooted at <code>vertex</code>, 
 spanning only the positions contained in <code>region</code>.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>genealogy</CODE> - Genealogy to clip <code>region</code> from.<DD><CODE>region</CODE> - Set of position to crop.<DD><CODE>vertex</CODE> - Vertex to start clipping from.
<DT><B>Returns:</B><DD>A new genealogy containing the clip.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE> - if <code>region</code> is not from the genealogy's snp domain.</DL>
</DD>
</DL>
<HR>

<A NAME="clone()"><!-- --></A><H3>
clone</H3>
<PRE>
public <A HREF="../../../../sanger/argml/graph/model/Genealogy.html" title="class in sanger.argml.graph.model">Genealogy</A> <B>clone</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>clone</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="haplotypes()"><!-- --></A><H3>
haplotypes</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A> <B>haplotypes</B>()</PRE>
<DL>
<DD>Reconstruct the haplotypes on all the leaves by tracing mutation events on the lineage.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>A collection of haplotypes for all leaves.</DL>
</DD>
</DL>
<HR>

<A NAME="haplotypes(sanger.math.set.NaturalSet)"><!-- --></A><H3>
haplotypes</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A> <B>haplotypes</B>(<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region)
                                throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD>Reconstruct the haplotypes in <code>region</code> on all the leaves by tracing mutation events on the lineage.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>region</CODE> - Set of SNPs to trace mutations for.
<DT><B>Returns:</B><DD>A collection of haplotypes for all leaves.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE> - if region is not defined over the genealogy's domain.</DL>
</DD>
</DL>
<HR>

<A NAME="mutations()"><!-- --></A><H3>
mutations</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A> <B>mutations</B>()</PRE>
<DL>
<DD>Trace all mutation events on the lineage leading to all the leaves. 
 The returned value contains SNP locations that have mutated since the GMRCA.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>A collection of mutation sets for all leaves.</DL>
</DD>
</DL>
<HR>

<A NAME="mutations(sanger.math.set.NaturalSet)"><!-- --></A><H3>
mutations</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A> <B>mutations</B>(<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region)
                               throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD>Trace all mutation events in <code>region</code> on the lineage leading to all the leaves. 
 The returned value contains SNP locations that have mutated since the GMRCA.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>region</CODE> - Set of SNPs to trace mutations for.
<DT><B>Returns:</B><DD>A collection of mutation sets for all leaves.
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE> - if region is not defined over the genealogy's domain.</DL>
</DD>
</DL>
<HR>

<A NAME="sort()"><!-- --></A><H3>
sort</H3>
<PRE>
public void <B>sort</B>()</PRE>
<DL>
<DD>Sort the vertex and edge collections according to their natural order.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>toString</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="snpDomain()"><!-- --></A><H3>
snpDomain</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A> <B>snpDomain</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="haplotypeDomain()"><!-- --></A><H3>
haplotypeDomain</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A> <B>haplotypeDomain</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="isTree()"><!-- --></A><H3>
isTree</H3>
<PRE>
public boolean <B>isTree</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>True if the genealogy has no recombination events.</DL>
</DD>
</DL>
<HR>

<A NAME="leaves()"><!-- --></A><H3>
leaves</H3>
<PRE>
public java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A>&gt; <B>leaves</B>()</PRE>
<DL>
<DD>Return a list of vertices with zero <A HREF="../../../../sanger/argml/graph/model/Vertex.html#outDegree()"><CODE>out degree</CODE></A>.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the leaf vertices</DL>
</DD>
</DL>
<HR>

<A NAME="vertices()"><!-- --></A><H3>
vertices</H3>
<PRE>
public java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A>&gt; <B>vertices</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the vertices</DL>
</DD>
</DL>
<HR>

<A NAME="edges()"><!-- --></A><H3>
edges</H3>
<PRE>
public java.util.LinkedList&lt;<A HREF="../../../../sanger/argml/graph/model/Edge.html" title="class in sanger.argml.graph.model">Edge</A>&gt; <B>edges</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the edges</DL>
</DD>
</DL>
<HR>

<A NAME="ancestralSequence()"><!-- --></A><H3>
ancestralSequence</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A> <B>ancestralSequence</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the ancestralSequence</DL>
</DD>
</DL>
<HR>

<A NAME="setAncestralSequence(sanger.math.set.NaturalSet)"><!-- --></A><H3>
setAncestralSequence</H3>
<PRE>
public void <B>setAncestralSequence</B>(<A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;ancestralSequence)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>ancestralSequence</CODE> - the ancestralSequence to set</DL>
</DD>
</DL>
<HR>

<A NAME="gmrca()"><!-- --></A><H3>
gmrca</H3>
<PRE>
public <A HREF="../../../../sanger/argml/graph/model/Vertex.html" title="class in sanger.argml.graph.model">Vertex</A> <B>gmrca</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the grand most recent common ancestor</DL>
</DD>
</DL>
<HR>

<A NAME="reportEmptyRegions()"><!-- --></A><H3>
reportEmptyRegions</H3>
<PRE>
public void <B>reportEmptyRegions</B>()</PRE>
<DL>
<DD>Report edges with empty active region to enviroment log.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="print(sanger.argml.io.TextOutput)"><!-- --></A><H3>
print</H3>
<PRE>
public void <B>print</B>(<A HREF="../../../../sanger/argml/io/TextOutput.html" title="class in sanger.argml.io">TextOutput</A>&nbsp;printer)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printDot2(sanger.argml.io.TextOutput)"><!-- --></A><H3>
printDot2</H3>
<PRE>
public void <B>printDot2</B>(<A HREF="../../../../sanger/argml/io/TextOutput.html" title="class in sanger.argml.io">TextOutput</A>&nbsp;printer)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="printDot1(sanger.argml.io.TextOutput)"><!-- --></A><H3>
printDot1</H3>
<PRE>
public void <B>printDot1</B>(<A HREF="../../../../sanger/argml/io/TextOutput.html" title="class in sanger.argml.io">TextOutput</A>&nbsp;printer)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../sanger/argml/graph/model/Edge.html" title="class in sanger.argml.graph.model"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../sanger/argml/graph/model/GenealogyFactory.html" title="class in sanger.argml.graph.model"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?sanger/argml/graph/model/Genealogy.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Genealogy.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>
Copyright (c) 2008 Genome Research Ltd.
</BODY>
</HTML>
