<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_33) on Tue Aug 14 12:30:35 EDT 2012 -->
<TITLE>
HaplotypeSet (${Name} API)
</TITLE>

<META NAME="date" CONTENT="2012-08-14">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="HaplotypeSet (${Name} API)";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../sanger/argml/graph/model/HaplotypeReader.HaplotypeIterator.html" title="class in sanger.argml.graph.model"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.HaplotypeIterator.html" title="class in sanger.argml.graph.model"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?sanger/argml/graph/model/HaplotypeSet.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="HaplotypeSet.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;<A HREF="#nested_class_summary">NESTED</A>&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
sanger.argml.graph.model</FONT>
<BR>
Class HaplotypeSet</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><A HREF="../../../../sanger/argml/environment/Environmental.html" title="class in sanger.argml.environment">sanger.argml.environment.Environmental</A>
      <IMG SRC="../../../../resources/inherit.gif" ALT="extended by "><B>sanger.argml.graph.model.HaplotypeSet</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.lang.Iterable&lt;<A HREF="../../../../sanger/argml/graph/model/Haplotype.html" title="class in sanger.argml.graph.model">Haplotype</A>&gt;</DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>HaplotypeSet</B><DT>extends <A HREF="../../../../sanger/argml/environment/Environmental.html" title="class in sanger.argml.environment">Environmental</A><DT>implements java.lang.Iterable&lt;<A HREF="../../../../sanger/argml/graph/model/Haplotype.html" title="class in sanger.argml.graph.model">Haplotype</A>&gt;</DL>
</PRE>

<P>
<HR>

<P>
<!-- ======== NESTED CLASS SUMMARY ======== -->

<A NAME="nested_class_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Nested Class Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;class</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.HaplotypeIterator.html" title="class in sanger.argml.graph.model">HaplotypeSet.HaplotypeIterator</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Iterates over <A HREF="../../../../sanger/argml/graph/model/Haplotype.html" title="class in sanger.argml.graph.model"><CODE>haplotypes</CODE></A> in a <A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model"><CODE>HaplotypeSet</CODE></A>.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#HaplotypeSet(sanger.argml.environment.Environment, sanger.math.set.NaturalDomain, sanger.math.set.NaturalDomain, int[])">HaplotypeSet</A></B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
             <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;haplotypeDomain,
             <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;snpDomain,
             int[]&nbsp;markerPositions)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a new empty haplotype set.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#add(sanger.argml.graph.model.Haplotype)">add</A></B>(<A HREF="../../../../sanger/argml/graph/model/Haplotype.html" title="class in sanger.argml.graph.model">Haplotype</A>&nbsp;haplotype)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#basePairDomain()">basePairDomain</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#clipHaplotypeDomain(sanger.argml.graph.model.HaplotypeSet, sanger.math.set.NaturalDomain)">clipHaplotypeDomain</A></B>(<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A>&nbsp;source,
                    <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;domain)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#clipSnpDomain(sanger.argml.graph.model.HaplotypeSet, sanger.math.set.NaturalSet)">clipSnpDomain</A></B>(<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A>&nbsp;source,
              <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#coordinate(int)">coordinate</A></B>(int&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#haplotypeDomain()">haplotypeDomain</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#haplotypes()">haplotypes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#hasMissing()">hasMissing</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Iterator&lt;<A HREF="../../../../sanger/argml/graph/model/Haplotype.html" title="class in sanger.argml.graph.model">Haplotype</A>&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#iterator()">iterator</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#marginalAlleleFrequencies()">marginalAlleleFrequencies</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#marginalAlleleFrequency(int)">marginalAlleleFrequency</A></B>(int&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#markerPositions()">markerPositions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#minorAlleleFrequencies()">minorAlleleFrequencies</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#minorAlleleFrequency(int)">minorAlleleFrequency</A></B>(int&nbsp;i)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#readFastPHASE(sanger.argml.environment.Environment, sanger.argml.io.TextInput)">readFastPHASE</A></B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
              <A HREF="../../../../sanger/argml/io/TextInput.html" title="class in sanger.argml.io">TextInput</A>&nbsp;fpin)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#readFastPHASE(sanger.argml.environment.Environment, sanger.argml.io.TextInput, sanger.argml.io.TextInput)">readFastPHASE</A></B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
              <A HREF="../../../../sanger/argml/io/TextInput.html" title="class in sanger.argml.io">TextInput</A>&nbsp;fpin,
              <A HREF="../../../../sanger/argml/io/TextInput.html" title="class in sanger.argml.io">TextInput</A>&nbsp;fpout)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#rsquare(int, int)">rsquare</A></B>(int&nbsp;i,
        int&nbsp;j)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#snpDomain()">snpDomain</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html#write(sanger.argml.io.TextOutput)">write</A></B>(<A HREF="../../../../sanger/argml/io/TextOutput.html" title="class in sanger.argml.io">TextOutput</A>&nbsp;out)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_sanger.argml.environment.Environmental"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class sanger.argml.environment.<A HREF="../../../../sanger/argml/environment/Environmental.html" title="class in sanger.argml.environment">Environmental</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../../sanger/argml/environment/Environmental.html#env()">env</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="HaplotypeSet(sanger.argml.environment.Environment, sanger.math.set.NaturalDomain, sanger.math.set.NaturalDomain, int[])"><!-- --></A><H3>
HaplotypeSet</H3>
<PRE>
public <B>HaplotypeSet</B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
                    <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;haplotypeDomain,
                    <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;snpDomain,
                    int[]&nbsp;markerPositions)</PRE>
<DL>
<DD>Constructs a new empty haplotype set.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>env</CODE> - The runtime enviroment<DD><CODE>haplotypeDomain</CODE> - The haplotype domain<DD><CODE>snpDomain</CODE> - The snp domain.<DD><CODE>markerPositions</CODE> - marker positions on the chromosome coordinates.</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="haplotypes()"><!-- --></A><H3>
haplotypes</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalSetCollection.html" title="class in sanger.math.set">NaturalSetCollection</A> <B>haplotypes</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="markerPositions()"><!-- --></A><H3>
markerPositions</H3>
<PRE>
public int[] <B>markerPositions</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="add(sanger.argml.graph.model.Haplotype)"><!-- --></A><H3>
add</H3>
<PRE>
public void <B>add</B>(<A HREF="../../../../sanger/argml/graph/model/Haplotype.html" title="class in sanger.argml.graph.model">Haplotype</A>&nbsp;haplotype)
         throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="clipSnpDomain(sanger.argml.graph.model.HaplotypeSet, sanger.math.set.NaturalSet)"><!-- --></A><H3>
clipSnpDomain</H3>
<PRE>
public static <A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A> <B>clipSnpDomain</B>(<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A>&nbsp;source,
                                         <A HREF="../../../../sanger/math/set/NaturalSet.html" title="class in sanger.math.set">NaturalSet</A>&nbsp;region)
                                  throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="clipHaplotypeDomain(sanger.argml.graph.model.HaplotypeSet, sanger.math.set.NaturalDomain)"><!-- --></A><H3>
clipHaplotypeDomain</H3>
<PRE>
public static <A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A> <B>clipHaplotypeDomain</B>(<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A>&nbsp;source,
                                               <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A>&nbsp;domain)
                                        throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="readFastPHASE(sanger.argml.environment.Environment, sanger.argml.io.TextInput, sanger.argml.io.TextInput)"><!-- --></A><H3>
readFastPHASE</H3>
<PRE>
public static <A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A> <B>readFastPHASE</B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
                                         <A HREF="../../../../sanger/argml/io/TextInput.html" title="class in sanger.argml.io">TextInput</A>&nbsp;fpin,
                                         <A HREF="../../../../sanger/argml/io/TextInput.html" title="class in sanger.argml.io">TextInput</A>&nbsp;fpout)
                                  throws java.io.IOException,
                                         <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A>,
                                         <A HREF="../../../../sanger/argml/graph/model/TooManyAllelesExcpetion.html" title="class in sanger.argml.graph.model">TooManyAllelesExcpetion</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE>
<DD><CODE><A HREF="../../../../sanger/argml/graph/model/TooManyAllelesExcpetion.html" title="class in sanger.argml.graph.model">TooManyAllelesExcpetion</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="readFastPHASE(sanger.argml.environment.Environment, sanger.argml.io.TextInput)"><!-- --></A><H3>
readFastPHASE</H3>
<PRE>
public static <A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.html" title="class in sanger.argml.graph.model">HaplotypeSet</A> <B>readFastPHASE</B>(<A HREF="../../../../sanger/argml/environment/Environment.html" title="class in sanger.argml.environment">Environment</A>&nbsp;env,
                                         <A HREF="../../../../sanger/argml/io/TextInput.html" title="class in sanger.argml.io">TextInput</A>&nbsp;fpin)
                                  throws java.io.IOException,
                                         <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A>,
                                         <A HREF="../../../../sanger/argml/graph/model/TooManyAllelesExcpetion.html" title="class in sanger.argml.graph.model">TooManyAllelesExcpetion</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE>
<DD><CODE><A HREF="../../../../sanger/argml/graph/model/TooManyAllelesExcpetion.html" title="class in sanger.argml.graph.model">TooManyAllelesExcpetion</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="write(sanger.argml.io.TextOutput)"><!-- --></A><H3>
write</H3>
<PRE>
public void <B>write</B>(<A HREF="../../../../sanger/argml/io/TextOutput.html" title="class in sanger.argml.io">TextOutput</A>&nbsp;out)
           throws java.io.IOException</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.io.IOException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>toString</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="iterator()"><!-- --></A><H3>
iterator</H3>
<PRE>
public java.util.Iterator&lt;<A HREF="../../../../sanger/argml/graph/model/Haplotype.html" title="class in sanger.argml.graph.model">Haplotype</A>&gt; <B>iterator</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE>iterator</CODE> in interface <CODE>java.lang.Iterable&lt;<A HREF="../../../../sanger/argml/graph/model/Haplotype.html" title="class in sanger.argml.graph.model">Haplotype</A>&gt;</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="basePairDomain()"><!-- --></A><H3>
basePairDomain</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A> <B>basePairDomain</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the basePairDomain</DL>
</DD>
</DL>
<HR>

<A NAME="haplotypeDomain()"><!-- --></A><H3>
haplotypeDomain</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A> <B>haplotypeDomain</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the haplotypeDomain</DL>
</DD>
</DL>
<HR>

<A NAME="snpDomain()"><!-- --></A><H3>
snpDomain</H3>
<PRE>
public <A HREF="../../../../sanger/math/set/NaturalDomain.html" title="class in sanger.math.set">NaturalDomain</A> <B>snpDomain</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the snpDomain</DL>
</DD>
</DL>
<HR>

<A NAME="hasMissing()"><!-- --></A><H3>
hasMissing</H3>
<PRE>
public boolean <B>hasMissing</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="coordinate(int)"><!-- --></A><H3>
coordinate</H3>
<PRE>
public int <B>coordinate</B>(int&nbsp;i)
               throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="marginalAlleleFrequencies()"><!-- --></A><H3>
marginalAlleleFrequencies</H3>
<PRE>
public double[] <B>marginalAlleleFrequencies</B>()
                                   throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="minorAlleleFrequencies()"><!-- --></A><H3>
minorAlleleFrequencies</H3>
<PRE>
public double[] <B>minorAlleleFrequencies</B>()
                                throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="marginalAlleleFrequency(int)"><!-- --></A><H3>
marginalAlleleFrequency</H3>
<PRE>
public double <B>marginalAlleleFrequency</B>(int&nbsp;i)
                               throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="minorAlleleFrequency(int)"><!-- --></A><H3>
minorAlleleFrequency</H3>
<PRE>
public double <B>minorAlleleFrequency</B>(int&nbsp;i)
                            throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="rsquare(int, int)"><!-- --></A><H3>
rsquare</H3>
<PRE>
public double <B>rsquare</B>(int&nbsp;i,
                      int&nbsp;j)
               throws <A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../../sanger/math/set/NaturalSetException.html" title="class in sanger.math.set">NaturalSetException</A></CODE></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../../sanger/argml/graph/model/HaplotypeReader.HaplotypeIterator.html" title="class in sanger.argml.graph.model"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../../sanger/argml/graph/model/HaplotypeSet.HaplotypeIterator.html" title="class in sanger.argml.graph.model"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../../index.html?sanger/argml/graph/model/HaplotypeSet.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="HaplotypeSet.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;<A HREF="#nested_class_summary">NESTED</A>&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>
Copyright (c) 2008 Genome Research Ltd.
</BODY>
</HTML>
